File: erlang/lectures/asynchronous/cell.erl
-module(cell).

-export([start/1, read/1, write/2]).

start(V) ->
   %% things to do in mother process
   {cell, spawn_link(fun() -> init(V) end)}.

read({cell, C}) -> 
    C ! {read, self()},
    receive 
	{value, V} ->
	    V
    end.

write({cell, C}, W) -> 
    C ! {write, W, self()},
    receive 
	ok ->
	    ok
    end.

init(V) ->
   %% things to do in the child process
   cell(V).

cell(V) ->
  receive 
      {read, Pid} ->
	  Pid ! {value, V},
	  cell(V);
      {write, W, Pid} ->
	  Pid ! ok,
	  cell(W)
  end.

File: lectures/asynchronous/src/cell.ex
defmodule Cell do

  def start(val) do       ## things to do in mother process
    {:cell, spawn_link(fn() -> init(val) end)}
  end

  def remote(node, val) do
    {:cell, Node.spawn_link(node, fn() -> init(val) end)}
  end
  
  
  def read({:cell, cell}) do
    send(cell , {:read, self()})
    receive do
      {:value, v} ->
	v
    end
  end

  def write({:cell, cell}, val) do
    send(cell, {:write, val, self()})
    receive do
    :ok ->
      :ok
    end
  end

  def quit({:cell, cell}) do
    send(cell, :quit)
  end  
  
  def init(val) do        ## things to do in the child process
    cell(val)
  end

  def cell(v) do
    receive do
      {:read, pid} ->
	send(pid, {:value, v})
	cell(v)
      {:write, w, pid} ->
	send(pid, :ok)
	cell(w)
      :quit ->
	:ok
    end
  end

end
  
